<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#План-работы:" data-toc-modified-id="План-работы:-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>План работы:</a></span></li><li><span><a href="#Загрузка-и-изучение-данных" data-toc-modified-id="Загрузка-и-изучение-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Загрузка и изучение данных</a></span><ul class="toc-item"><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Анализ-данных" data-toc-modified-id="Анализ-данных-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Анализ данных</a></span></li><li><span><a href="#Предобработка" data-toc-modified-id="Предобработка-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Предобработка</a></span></li><li><span><a href="#Объединение-таблиц" data-toc-modified-id="Объединение-таблиц-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Объединение таблиц</a></span></li></ul></li><li><span><a href="#Создание-таргета-и-фича-инжиниринг" data-toc-modified-id="Создание-таргета-и-фича-инжиниринг-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Создание таргета и фича-инжиниринг</a></span><ul class="toc-item"><li><span><a href="#Создание-таргета" data-toc-modified-id="Создание-таргета-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Создание таргета</a></span></li><li><span><a href="#Фича-инжиниринг" data-toc-modified-id="Фича-инжиниринг-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Фича-инжиниринг</a></span></li><li><span><a href="#Эмбеддинг" data-toc-modified-id="Эмбеддинг-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Эмбеддинг</a></span></li></ul></li><li><span><a href="#Создание-выборок-и-обучение-модели" data-toc-modified-id="Создание-выборок-и-обучение-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Создание выборок и обучение модели</a></span><ul class="toc-item"><li><span><a href="#Создание-выборок" data-toc-modified-id="Создание-выборок-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Создание выборок</a></span></li><li><span><a href="#Обучение-модели" data-toc-modified-id="Обучение-модели-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Обучение модели</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Выгрузка-модели" data-toc-modified-id="Выгрузка-модели-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Выгрузка модели</a></span></li></ul></div>

# **Описание проекта. Матчинг карточек товара.**
**Заказчик**

ООО «ПРОСЕПТ» — российская производственная компания, специализирующаяся на выпуске профессиональной химии. В своей работе используют опыт ведущих мировых производителей и сырье крупнейших химических концернов. Производство и логистический центр расположены в непосредственной близости от Санкт-Петербурга, откуда продукция компании поставляется во все регионы России.

Сайт: https://prosept.ru/

**Описание проекта**

Заказчик производит несколько сотен различных товаров бытовой и промышленной химии, а затем продаёт эти товары через дилеров. Дилеры, в свою очередь, занимаются розничной продажей товаров в крупных сетях магазинов и на онлайн площадках.
Для оценки ситуации,  управления ценами и  бизнесом в целом, заказчик периодически собирает информацию о том, как дилеры продают их товар. Для этого они парсят сайты дилеров, а затем сопоставляют товары и цены.
Зачастую описание товаров на сайтах дилеров отличаются от того описания, что даёт заказчик. Например, могут добавляться новый слова (“универсальный”, “эффективный”), объём (0.6 л -> 600 мл). Поэтому сопоставление товаров дилеров с товарами производителя делается вручную.

**Цель этого проекта** - разработка решения, которое отчасти автоматизирует процесс сопоставления товаров. Основная идея - предлагать несколько товаров заказчика, которые с наибольшей вероятностью соответствуют размечаемому товару дилера.
Предлагается реализовать это решение, как онлайн сервис, открываемый в веб-браузере. Выбор наиболее вероятных подсказок делается методами машинного обучения

**Предоставление вариантов соответствия**:
Система должна использовать рекомендательную модель, созданную ML специалистами, для предоставления наиболее вероятных вариантов соответствия товара.
Для каждого товара система должна предложить несколько (настраиваемое количество) вариантов соответствия.

**Описание данных**
Заказчик предоставил несколько таблиц (дамп БД), содержащих необходимые данные:

1.	marketing_dealer - список дилеров;
2.	marketing_dealerprice - результат работы парсера площадок дилеров:
  - product_key - уникальный номер позиции;
  - price - цена;
  - product_url - адрес страницы, откуда собраны данные;
  - product_name - заголовок продаваемого товара;
  - date - дата получения информации;
  - dealer_id - идентификатор дилера (внешний ключ к marketing_dealer)
3.	marketing_product - список товаров, которые производит и распространяет заказчик;
  - article - артикул товара;
  - ean_13 - код товара (см. EAN 13)
  - name - название товара;
  - cost - стоимость;
  - min_recommended_price - рекомендованная минимальная цена;
  - recommended_price - рекомендованная цена;
  - category_id - категория товара;
  - ozon_name - названиет товара на Озоне;
  - name_1c - название товара в 1C;
  - Wb_name - название товара на Wildberries;
  - ozon_article - описание для Озон;
  - wb_article - артикул для Wildberries;
  - ym_article - артикул для Яндекс.Маркета;

4.	marketing_productdealerkey - таблица матчинга товаров заказчика и товаров дилеров
  - key - внешний ключ к marketing_dealerprice
  - product_id - внешний ключ к marketing_product
  - dealer_id - внешний ключ к marketing_dealer


## План работы:
1. Загрузка и изучение данных;  
2. Предобработка данных;
3. Выбор и создание признаков;
4. Разделение на выборки;
5. Обучение и выбор модели;
6. Тестирование модели;
7. Отдать модель в backend.

## Загрузка и изучение данных  
Загружаем исходные данные

### Импорт библиотек

In [1]:
!pip install fuzzywuzzy -q
!pip install python-Levenshtein -q

In [2]:
import re

import numpy as np
import pandas as pd
import pickle
import torch
import warnings


from fuzzywuzzy import fuzz
from lightgbm import LGBMClassifier
from sklearn.model_selection import (
    GridSearchCV,
    train_test_split    
)
from sklearn.metrics import precision_score
from transformers import AutoTokenizer, AutoModel, logging


logging.set_verbosity_error()
warnings.filterwarnings('ignore')


RANDOM_STATE=12345

### Загрузка данных

In [3]:
dealer = pd.read_csv('C:\\data\\marketing_dealer.csv', sep=';', index_col='id')
dealer_price = pd.read_csv('C:\\data\\marketing_dealerprice.csv', sep=';', index_col='id')
product = pd.read_csv('C:\\data\\marketing_product.csv', sep=';', index_col='id')
product_dealer_key = pd.read_csv('C:\\data\\marketing_productdealerkey.csv', sep=';', index_col='id')
product.drop(['Unnamed: 0'], axis=1, inplace=True)

In [4]:
dealer.sort_index(inplace=True)
dealer_price.sort_index(inplace=True)
product.sort_index(inplace=True)
product_dealer_key.sort_index(inplace=True)

### Анализ данных

**Dealer**

In [5]:
dealer.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18 entries, 1 to 18
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   name    18 non-null     object
dtypes: object(1)
memory usage: 288.0+ bytes


In [6]:
dealer

name
id                          
1               Moi_vibor_WB
2                      Akson
3                      Bafus
4                  Baucenter
5                  Castorama
6                   Cubatora
7                      Komus
8               Leroy_Merlin
9                  Megastroy
10               OnlineTrade
11                 Petrovich
12                     sdvor
13                  simaLand
14                    VegosM
15            Vse_instrumeni
16                     Vimos
17           Unicleaner_OZON
18  Мasterstroy_spb_OZON\r\n

Всего 18 дилеров, дубликатов нет

**Dealer_price**

In [7]:
dealer_price.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20416 entries, 2 to 20570
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   product_key   20416 non-null  object 
 1   price         20416 non-null  float64
 2   product_url   20182 non-null  object 
 3   product_name  20416 non-null  object 
 4   date          20416 non-null  object 
 5   dealer_id     20416 non-null  int64  
dtypes: float64(1), int64(1), object(4)
memory usage: 1.1+ MB


Есть пропуски в признаке с ссылкой, надо проверить на дубликаты

In [8]:
dealer_price

product_key   price                                        product_url  \
id                                                                             
2          546227   233.0  https://akson.ru//p/sredstvo_universalnoe_pros...   
3          546408   175.0  https://akson.ru//p/kontsentrat_prosept_multip...   
4          546234   285.0  https://akson.ru//p/sredstvo_dlya_chistki_lyus...   
5          651258   362.0  https://akson.ru//p/udalitel_rzhavchiny_prosep...   
6          546355   205.0  https://akson.ru//p/sredstvo_moyushchee_dlya_b...   
...           ...     ...                                                ...   
20566   534517451   823.0              https://www.ozon.ru/product/534517451   
20567   530308963   418.0              https://www.ozon.ru/product/530308963   
20568   531730388  2986.0              https://www.ozon.ru/product/531730388   
20569  1090913025   579.0             https://www.ozon.ru/product/1090913025   
20570  1077090171  2623.0             https://www.ozon.ru/product/1077090171   

                                            product_name        date  \
id                                                                     
2      Средство универсальное Prosept Universal Spray...  2023-07-11   
3      Концентрат Prosept Multipower для мытья полов,...  2023-07-11   
4      Средство для чистки люстр Prosept Universal An...  2023-07-11   
5      Удалитель ржавчины PROSEPT RUST REMOVER 0,5л 0...  2023-07-11   
6      Средство моющее для бани и сауны Prosept Multi...  2023-07-11   
...                                                  ...         ...   
20566  Огнебиозащита для древесины PROSEPT 2 группа, ...  2023-07-31   
20567       Антисептик многофункциональный ФБС, ГОСТ 5 л  2023-07-31   
20568       Средство для удаления ржавчины PROSEPT, 1 шт  2023-07-31   
20569  Герметик акриловый межшовный для деревянных ко...  2023-07-31   
20570  Краска-грунт фасадная для плит OSB Proff 3 в 1...  2023-07-31   

       dealer_id  
id                
2              2  
3              2  
4              2  
5              2  
6              2  
...          ...  
20566         18  
20567         18  
20568         18  
20569         18  
20570         18  

[20416 rows x 6 columns]

In [9]:
dealer_price[dealer_price['product_key'].duplicated()].shape

(18451, 6)

Есть пропуски в product_url, но они не важны. Очень много дубликатов - 18451 из 20363. Отсортирую по дате, оставлю самые актуальные записи.
Оставлю признаки - product_key (для матчинга) и product_name.

product_key переименую в key (для джоина).

**Product_dealer_key**

In [10]:
product_dealer_key.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1700 entries, 1 to 2023
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   key         1700 non-null   object
 1   dealer_id   1700 non-null   int64 
 2   product_id  1700 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 53.1+ KB


In [11]:
product_dealer_key

key  dealer_id  product_id
id                                    
1        546227          2          12
2        651265          2         106
3        546257          2         200
4        546408          2          38
5        651258          2         403
...         ...        ...         ...
2019  530017190         18         267
2020  530139882         18         286
2021  534659036         18         129
2022  898350801         18           1
2023  534666258         18         128

[1700 rows x 3 columns]

In [12]:
product_dealer_key.describe()

dealer_id   product_id
count  1700.000000  1700.000000
mean     10.148824   223.564118
std       5.830717   132.341778
min       1.000000     1.000000
25%       3.000000   110.000000
50%      10.000000   239.500000
75%      15.000000   334.250000
max      18.000000   479.000000

In [13]:
product_dealer_key.duplicated().sum()

0

Пропусков нет.

**Product**

In [14]:
product.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 496 entries, 1 to 513
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   article            496 non-null    object 
 1   ean_13             464 non-null    float64
 2   name               494 non-null    object 
 3   cost               491 non-null    float64
 4   recommended_price  491 non-null    float64
 5   category_id        447 non-null    float64
 6   ozon_name          458 non-null    object 
 7   name_1c            485 non-null    object 
 8   wb_name            455 non-null    object 
 9   ozon_article       365 non-null    float64
 10  wb_article         340 non-null    float64
 11  ym_article         337 non-null    object 
 12  wb_article_td      32 non-null     object 
dtypes: float64(6), object(7)
memory usage: 54.2+ KB


In [15]:
product

article        ean_13                                               name  \
id                                                                             
1    123-02  4.610093e+12  Средство для удаления наклеек Duty Universal /...   
2     242-6  4.610093e+12      Антигололед - 32 PROSEPTготовый состав / 6 кг   
3    242-12           NaN     Антигололед - 32 PROSEPTготовый состав / 12 кг   
4     104-1  4.680008e+12  Универсальное моющее средствоUniversal Prof  к...   
5     104-5  4.680008e+12  Универсальное моющее средствоUniversal Prof  к...   
..      ...           ...                                                ...   
509  М004-2  4.610093e+12  Набор для посудомоечной посуды PROSEPT, 2 сред...   
510  М005-2  4.610093e+12  Набор "Антиплесень" PROSEPT (удалитель плесени...   
511  М006-2  4.610093e+12  Набор для удаления загрязнений и наклеек PROSE...   
512  М009-2  4.610093e+12  Средство для мытья полов PROSEPT Multipower "В...   
513  М011-2  4.610093e+12  Средство для мытья полов PROSEPT Multipower "П...   

       cost  recommended_price  category_id  \
id                                            
1    189.00              443.0         43.0   
2    220.39              515.0          NaN   
3    460.16             1075.0          NaN   
4    150.00              352.0         56.0   
5    601.76             1405.0         56.0   
..      ...                ...          ...   
509  322.00              752.0          NaN   
510  354.00              827.0          NaN   
511  284.00              664.0          NaN   
512  156.00              364.0          NaN   
513  156.00              364.0          NaN   

                                             ozon_name  \
id                                                       
1    Средство для удаления клейкой ленты, клея, нак...   
2                                                  NaN   
3                                                  NaN   
4    Профессиональное универсальное моющее средство...   
5    Профессиональное универсальное моющее средство...   
..                                                 ...   
509  Набор для посудомоечной посуды PROSEPT, 2 сред...   
510  Набор "Антиплесень" PROSEPT (удалитель плесени...   
511  Набор для удаления загрязнений и наклеек PROSE...   
512  Средство для мытья полов PROSEPT Multipower "В...   
513  Средство для мытья полов PROSEPT Multipower "П...   

                                               name_1c  \
id                                                       
1    Средство для удаления клейкой ленты, клея, нак...   
2        Антигололед - 32 PROSEPTготовый состав / 6 кг   
3       Антигололед - 32 PROSEPTготовый состав / 12 кг   
4    Профессиональное универсальное моющее средство...   
5    Профессиональное универсальное моющее средство...   
..                                                 ...   
509  Набор для посудомоечной машины - 1 PROSEPT, 2 ...   
510  Набор "Антиплесень" PROSEPT (удалитель плесени...   
511   Набор для удаления загрязнений и наклеек PROSEPT   
512  Набор средств для мытья полов PROSEPT Multipow...   
513  Набор средств для мытья полов PROSEPT Multipow...   

                                               wb_name  ozon_article  \
id                                                                     
1    Средство для удаления клейкой ленты, клея, нак...   445949007.0   
2                                                  NaN           NaN   
3                                                  NaN           NaN   
4    Профессиональное универсальное моющее средство...   449915357.0   
5    Профессиональное универсальное моющее средство...   449915356.0   
..                                                 ...           ...   
509  Набор для посудомоечной посуды Splash, ополаск...   463671331.0   
510  Набор "Антиплесень" :удалитель плесени и защит...   463682015.0   
511  Набор для удаления загрязнений и наклеек , 2 с...   464002480.0   
512                                              

In [16]:
product.duplicated().sum()

0

Есть пропуски, дубликатов нет. Оставлю признаки - name и category_id. Пропуски в category_id заполню заглушкой 0. Сброшу индекс в признаки, чтобы заджоинить с таблицей матчинга и переименую.

### Предобработка

In [17]:
def re_name(name):
    """
    Функция предобработки текста.
    Разделяет некоторые слитные слова.
    мылоPROSEPT -> мыло PROSEPT
    """
    return re.sub(r'(?<=[а-яa-z])(?=[A-Z])|(?=[а-я])(?<=[A-Za-z])|(?<=[a-z])(?=[0-9])', ' ', name)

**dealer_price**

In [18]:
#сортировка дилерских товаров по дате, чтобы удалить дубликаты и оставить только актуальные данные
dealer_price['date'] = pd.to_datetime(dealer_price['date'])
dealer_price.sort_values('date', ascending=False, inplace=True)

In [19]:
dealer_price

product_key   price                                        product_url  \
id                                                                             
20570  1077090171  2623.0             https://www.ozon.ru/product/1077090171   
19395     1656319   219.0  https://www.onlinetrade.ru/catalogue/rastvorit...   
19384     1923011   119.0  https://www.onlinetrade.ru/catalogue/universal...   
19385     1923010   129.0  https://www.onlinetrade.ru/catalogue/sredstva_...   
19386     3201601   400.0  https://www.onlinetrade.ru/catalogue/propitki_...   
...           ...     ...                                                ...   
1232     15728509   495.0  https://www.vseinstrumenti.ru/product/antisept...   
1233     20633710  1667.0  https://www.vseinstrumenti.ru/product/antisept...   
1234     15728524   529.0  https://www.vseinstrumenti.ru/product/udalitel...   
1235     15728556   200.0  https://www.vseinstrumenti.ru/product/sredstvo...   
2          546227   233.0  https://akson.ru//p/sredstvo_universalnoe_pros...   

                                            product_name       date  dealer_id  
id                                                                              
20570  Краска-грунт фасадная для плит OSB Proff 3 в 1... 2023-07-31         18  
19395  Удалитель ржавчины PROSEPT RUST REMOVER, 500 м... 2023-07-31         10  
19384  Чистящий крем PROSEPT Cooky Universal Ледяная ... 2023-07-31         10  
19385  Чистящее средство PROSEPT Bath Acid Plus Цитру... 2023-07-31         10  
19386  Антисептик трудновымываемый PROSEPT ХМББ, ГОСТ... 2023-07-31         10  
...                                                  ...        ...        ...  
1232   Антисептик PROSEPT EXTERIOR для наружных работ... 2023-07-11         15  
1233   Антисептик PROSEPT ECO ULTRA невымываемый, кор... 2023-07-11         15  
1234   Удалитель плесени FUNGI CLEAN (концентрат) 1 л... 2023-07-11         15  
1235   Средство Optic Crystal для мытья стекол и зерк... 2023-07-11         15  
2      Средство универсальное Prosept Universal Spray... 2023-07-11          2  

[20416 rows x 6 columns]

In [20]:
dealer_price[dealer_price['product_key'].duplicated(keep=False)].sort_values(['product_key', 'date'])

product_key   price  \
id                                                                 
15476                                             382751   924.0   
20302                                             382751   924.0   
15490                                             382752   646.0   
20318                                             382752   646.0   
15491                                             382754   784.0   
...                                                  ...     ...   
11855  https://kub02.ru/catalog/prosept/ognebiozashch...  1763.0   
13186  https://kub02.ru/catalog/prosept/ognebiozashch...  1763.0   
14622  https://kub02.ru/catalog/prosept/ognebiozashch...  1763.0   
16053  https://kub02.ru/catalog/prosept/ognebiozashch...  1763.0   
19343  https://kub02.ru/catalog/prosept/ognebiozashch...  1763.0   

                                             product_url  \
id                                                         
15476  https://akson.ru//p/ognebiozaschita_prosept_pr...   
20302  https://akson.ru//p/ognebiozaschita_prosept_pr...   
15490  https://akson.ru//p/antiseptik_prosept_exterio...   
20318  https://akson.ru//p/antiseptik_prosept_exterio...   
15491  https://akson.ru//p/antiseptik_prosept_ultra_n...   
...                                                  ...   
11855  https://kub02.ru/catalog/prosept/ognebiozashch...   
13186  https://kub02.ru/catalog/prosept/ognebiozashch...   
14622  https://kub02.ru/catalog/prosept/ognebiozashch...   
16053  https://kub02.ru/catalog/prosept/ognebiozashch...   
19343  https://kub02.ru/catalog/prosept/ognebiozashch...   

                                            product_name       date  dealer_id  
id                                                                              
15476  Огнебиозащита PROSEPT prof 2 группа для наружн... 2023-07-26          2  
20302  Огнебиозащита PROSEPT prof 2 группа для наружн... 2023-07-31          2  
15490  Пропитка-антисептик PROSEPT Exterior трудновым... 2023-07-26          2  
20318  Пропитка-антисептик PROSEPT Exterior трудновым... 2023-07-31          2  
15491  Пропитка-антисептик PROSEPT Ultra невымываемая... 2023-07-26          2  
...                                                  ...        ...        ...  
11855  Огнебиозащита для древесины PROSEPT, 1 группа ... 2023-07-21          6  
13186  Огнебиозащита для древесины PROSEPT, 1 группа ... 2023-07-24          6  
14622  Огнебиозащита для древесины PROSEPT, 1 группа ... 2023-07-25          6  
16053  Огнебиозащита для древесины PROSEPT, 1 группа ... 2023-07-26          6  
19343  Огнебиозащита для древесины PROSEPT, 1 группа ... 2023-07-31          6  

[20363 rows x 6 columns]

In [21]:
dealer_price_new = dealer_price.copy().sort_values(['product_key', 'date']).drop(['date'], axis=1).drop_duplicates(keep='last')

In [22]:
dealer_price_new

product_key   price  \
id                                                                 
20284                                             382745  1771.0   
20302                                             382751   924.0   
20318                                             382752   646.0   
15492                                             382753   683.0   
20323                                             382754   784.0   
...                                                  ...     ...   
19344  https://kub02.ru/catalog/prosept/antiseptik_un...   620.0   
19336  https://kub02.ru/catalog/prosept/lak_dlya_bani...   531.0   
19337  https://kub02.ru/catalog/prosept/maslo_dlya_za...   418.0   
19343  https://kub02.ru/catalog/prosept/ognebiozashch...  1763.0   
19346  https://kub02.ru/catalog/prosept/ognebiozashch...   888.0   

                                             product_url  \
id                                                         
20284  https://akson.ru//p/ognebiozaschita_prosept_pr...   
20302  https://akson.ru//p/ognebiozaschita_prosept_pr...   
20318  https://akson.ru//p/antiseptik_prosept_exterio...   
15492  https://akson.ru//p/antiseptik_prosept_univers...   
20323  https://akson.ru//p/antiseptik_prosept_ultra_n...   
...                                                  ...   
19344  https://kub02.ru/catalog/prosept/antiseptik_un...   
19336  https://kub02.ru/catalog/prosept/lak_dlya_bani...   
19337  https://kub02.ru/catalog/prosept/maslo_dlya_za...   
19343  https://kub02.ru/catalog/prosept/ognebiozashch...   
19346  https://kub02.ru/catalog/prosept/ognebiozashch...   

                                            product_name  dealer_id  
id                                                                   
20284  Огнебиозащита PROSEPT prof 1 группа для наружн...          2  
20302  Огнебиозащита PROSEPT prof 2 группа для наружн...          2  
20318  Пропитка-антисептик PROSEPT Exterior трудновым...          2  
15492  Пропитка-антисептик PROSEPT Universal универса...          2  
20323  Пропитка-антисептик PROSEPT Ultra невымываемая...          2  
...                                                  ...        ...  
19344  Антисептик универсальный против грибка и плесе...          6  
19336  Лак для бани и сауны термостойкий, акриловый P...          6  
19337  Масло для защиты полков PROSEPT SAUNA OIL, гот...          6  
19343  Огнебиозащита для древесины PROSEPT, 1 группа ...          6  
19346  Огнебиозащита для древесины PROSEPT, 2 группа ...          6  

[3024 rows x 5 columns]

In [23]:
dealer_price_new.drop(['product_url', 'price', 'dealer_id'], axis=1, inplace=True)
dealer_price_new.sort_index(inplace=True)
dealer_price_new['product_name'] = dealer_price_new['product_name'].apply(re_name)
dealer_price_new = dealer_price_new.rename(columns={'product_key': 'key'})
dealer_price_new

key                                       product_name
id                                                                  
429    1001472268        Средстводлямытьяполов Prosept Multipower 1л
433    1001472256  Средстводляудаленияплесени Prosept Fungi Clean...
452    1001472273      Чистящеесредство Prosept Universal Spray,0,5л
529    1001472253  Средстводляудаленияклея Prosept Duty Universal...
919        132643  Средство для удаления цемента Prosept Cement C...
...           ...                                                ...
20566   534517451  Огнебиозащита для древесины PROSEPT 2 группа, ...
20567   530308963       Антисептик многофункциональный ФБС, ГОСТ 5 л
20568   531730388       Средство для удаления ржавчины PROSEPT, 1 шт
20569  1090913025  Герметик акриловый межшовный для деревянных ко...
20570  1077090171  Краска-грунт фасадная для плит OSB Proff 3 в 1...

[3024 rows x 2 columns]

**product**

In [24]:
product

article        ean_13                                               name  \
id                                                                             
1    123-02  4.610093e+12  Средство для удаления наклеек Duty Universal /...   
2     242-6  4.610093e+12      Антигололед - 32 PROSEPTготовый состав / 6 кг   
3    242-12           NaN     Антигололед - 32 PROSEPTготовый состав / 12 кг   
4     104-1  4.680008e+12  Универсальное моющее средствоUniversal Prof  к...   
5     104-5  4.680008e+12  Универсальное моющее средствоUniversal Prof  к...   
..      ...           ...                                                ...   
509  М004-2  4.610093e+12  Набор для посудомоечной посуды PROSEPT, 2 сред...   
510  М005-2  4.610093e+12  Набор "Антиплесень" PROSEPT (удалитель плесени...   
511  М006-2  4.610093e+12  Набор для удаления загрязнений и наклеек PROSE...   
512  М009-2  4.610093e+12  Средство для мытья полов PROSEPT Multipower "В...   
513  М011-2  4.610093e+12  Средство для мытья полов PROSEPT Multipower "П...   

       cost  recommended_price  category_id  \
id                                            
1    189.00              443.0         43.0   
2    220.39              515.0          NaN   
3    460.16             1075.0          NaN   
4    150.00              352.0         56.0   
5    601.76             1405.0         56.0   
..      ...                ...          ...   
509  322.00              752.0          NaN   
510  354.00              827.0          NaN   
511  284.00              664.0          NaN   
512  156.00              364.0          NaN   
513  156.00              364.0          NaN   

                                             ozon_name  \
id                                                       
1    Средство для удаления клейкой ленты, клея, нак...   
2                                                  NaN   
3                                                  NaN   
4    Профессиональное универсальное моющее средство...   
5    Профессиональное универсальное моющее средство...   
..                                                 ...   
509  Набор для посудомоечной посуды PROSEPT, 2 сред...   
510  Набор "Антиплесень" PROSEPT (удалитель плесени...   
511  Набор для удаления загрязнений и наклеек PROSE...   
512  Средство для мытья полов PROSEPT Multipower "В...   
513  Средство для мытья полов PROSEPT Multipower "П...   

                                               name_1c  \
id                                                       
1    Средство для удаления клейкой ленты, клея, нак...   
2        Антигололед - 32 PROSEPTготовый состав / 6 кг   
3       Антигололед - 32 PROSEPTготовый состав / 12 кг   
4    Профессиональное универсальное моющее средство...   
5    Профессиональное универсальное моющее средство...   
..                                                 ...   
509  Набор для посудомоечной машины - 1 PROSEPT, 2 ...   
510  Набор "Антиплесень" PROSEPT (удалитель плесени...   
511   Набор для удаления загрязнений и наклеек PROSEPT   
512  Набор средств для мытья полов PROSEPT Multipow...   
513  Набор средств для мытья полов PROSEPT Multipow...   

                                               wb_name  ozon_article  \
id                                                                     
1    Средство для удаления клейкой ленты, клея, нак...   445949007.0   
2                                                  NaN           NaN   
3                                                  NaN           NaN   
4    Профессиональное универсальное моющее средство...   449915357.0   
5    Профессиональное универсальное моющее средство...   449915356.0   
..                                                 ...           ...   
509  Набор для посудомоечной посуды Splash, ополаск...   463671331.0   
510  Набор "Антиплесень" :удалитель плесени и защит...   463682015.0   
511  Набор для удаления загрязнений и наклеек , 2 с...   464002480.0   
512                                              

In [25]:
product_dealer_key.max()['product_id']

479

In [26]:
product.describe()

ean_13         cost  recommended_price  category_id  \
count  4.640000e+02   491.000000         491.000000   447.000000   
mean   4.667050e+12   880.162709        1880.372790    34.342282   
std    2.719656e+10  1245.260825        2518.474824    10.584979   
min    4.610093e+12    44.000000           0.000000    20.000000   
25%    4.680008e+12   201.000000         436.000000    25.000000   
50%    4.680008e+12   409.000000         891.000000    33.000000   
75%    4.680008e+12   839.500000        1799.500000    41.000000   
max    4.680008e+12  8011.000000       14980.000000    57.000000   

       ozon_article    wb_article  
count  3.650000e+02  3.400000e+02  
mean   3.789162e+08  1.505817e+08  
std    1.289826e+08  2.506685e+06  
min    1.895227e+08  1.483674e+08  
25%    2.326860e+08  1.498110e+08  
50%    4.498035e+08  1.500330e+08  
75%    4.529848e+08  1.501234e+08  
max    5.776281e+08  1.702115e+08

In [27]:
#оставлю только те строки, которые есть в таблице для матчинга, оставлю признаки name и category_id
product_new = product.copy().loc[:product_dealer_key.max()['product_id']][['name', 'category_id']]
#пропуски в category_id заполню заглушкой 0
product_new['category_id'] = product_new['category_id'].fillna(0).astype(int)
product_new['name'] = product_new['name'].apply(re_name)

In [28]:
product_new

name  category_id
id                                                                 
1    Средство для удаления наклеек Duty Universal /...           43
2       Антигололед - 32 PROSEPT готовый состав / 6 кг            0
3      Антигололед - 32 PROSEPT готовый состав / 12 кг            0
4    Универсальное моющее средство Universal Prof  ...           56
5    Универсальное моющее средство Universal Prof  ...           56
..                                                 ...          ...
472  ОГНЕБИОЗАЩИТА для древесины 2 группа, красный ...           38
476  ОГНЕБИОЗАЩИТА для древесины 2 группа, бесцветн...           38
477       Набор для люстр Anti-Dust спрей, 0.5 л, 2 шт           40
478       Набор для люстр Anti-Dust спрей, 0.5 л, 3 шт           40
479  PROSEPT огнебиозащита для древесины с цветовой...            0

[462 rows x 2 columns]

In [29]:
#сбрасываю индексы, чтобы вывести id в признак и переименовываю для merge
product_new = product_new.reset_index().rename(columns={'id': 'product_id'})
product_new

product_id                                               name  \
0             1  Средство для удаления наклеек Duty Universal /...   
1             2     Антигололед - 32 PROSEPT готовый состав / 6 кг   
2             3    Антигололед - 32 PROSEPT готовый состав / 12 кг   
3             4  Универсальное моющее средство Universal Prof  ...   
4             5  Универсальное моющее средство Universal Prof  ...   
..          ...                                                ...   
457         472  ОГНЕБИОЗАЩИТА для древесины 2 группа, красный ...   
458         476  ОГНЕБИОЗАЩИТА для древесины 2 группа, бесцветн...   
459         477       Набор для люстр Anti-Dust спрей, 0.5 л, 2 шт   
460         478       Набор для люстр Anti-Dust спрей, 0.5 л, 3 шт   
461         479  PROSEPT огнебиозащита для древесины с цветовой...   

     category_id  
0             43  
1              0  
2              0  
3             56  
4             56  
..           ...  
457           38  
458           38  
459           40  
460           40  
461            0  

[462 rows x 3 columns]

### Объединение таблиц

In [30]:
product_dealer_key['key'] = product_dealer_key.key.astype(str)
dealer_price_new['key'] = dealer_price_new.key.astype(str)

In [31]:
df = product_dealer_key.merge(dealer_price_new, on='key', how='right')
df = df.merge(product_new, on='product_id', how='left')

In [32]:
df

key  dealer_id  product_id  \
0     1001472268        5.0        38.0   
1     1001472256        5.0       390.0   
2     1001472273        5.0        12.0   
3     1001472253        5.0       128.0   
4         132643       12.0       400.0   
...          ...        ...         ...   
3019   534517451       18.0       452.0   
3020   530308963       18.0       416.0   
3021   531730388       18.0       405.0   
3022  1090913025        NaN         NaN   
3023  1077090171        NaN         NaN   

                                           product_name  \
0           Средстводлямытьяполов Prosept Multipower 1л   
1     Средстводляудаленияплесени Prosept Fungi Clean...   
2         Чистящеесредство Prosept Universal Spray,0,5л   
3     Средстводляудаленияклея Prosept Duty Universal...   
4     Средство для удаления цемента Prosept Cement C...   
...                                                 ...   
3019  Огнебиозащита для древесины PROSEPT 2 группа, ...   
3020       Антисептик многофункциональный ФБС, ГОСТ 5 л   
3021       Средство для удаления ржавчины PROSEPT, 1 шт   
3022  Герметик акриловый межшовный для деревянных ко...   
3023  Краска-грунт фасадная для плит OSB Proff 3 в 1...   

                                                   name  category_id  
0     Концентрат для мытья полов MULTIPOWER с аромат...         50.0  
1     Удалитель плесени FUNGI CLEAN готовый состав /...         35.0  
2     Универсальное моющее и чистящее средство Unive...         56.0  
3     Средство для удаления клейкой ленты, клея, нак...         43.0  
4     Удалитель цемента CEMENT CLEANER готовый соста...         35.0  
...                                                 ...          ...  
3019  ОГНЕБИОЗАЩИТА для древесины 2 группа, красный ...         38.0  
3020     Антисептик многофункциональный ФБС, ГОСТ / 5 л         20.0  
3021  Удалитель ржавчины RUST REMOVER концентрат  1:...         35.0  
3022                                                NaN          NaN  
3023                                                NaN          NaN  

[3024 rows x 6 columns]

In [33]:
df.isna().sum()

key               0
dealer_id       423
product_id      423
product_name      0
name            423
category_id     423
dtype: int64

**Вывод:** В таблице `marketing_dealer`, всего 18 строк, дубликатов нет.

`dealer_price` содержала 20416 строк, пропуски в product_url, есть дубликаты (парсился несколько раз один и тот же продукт, но в разные дни), остаются признаки - `product_key` (для матчинга) и `product_name`. `product_key` переименую в `key`. После удаления дубликатов осталось 3024,

`product_dealer_key` имеет 1700 строк, дубликатов нет, на основе этой таблицы будет создан таргет - все эти связи имеют таргет 1, строки без пары будут иметь 0.

`product` содержит 496 строк, но есть пропуски в индексе - максимальный 513. Пропуски в `ean_13`, `name`, `cost`, `recommended_price`, `category_id`, `ozon_name`, `name_1c`, `wb_name`, `ozon_article`, `wb_article`, `ym_article`, `wb_article_td`. Оставлю признаки - `name` и `category_id`. Пропуски в `category_id` заполню заглушкой 0 и переведу в int. 

Таблицы `product`, `product_dealer_key` и `dealer_price` были объединены в одну, признаки - `key`, `dealer_id`, `product_id`, `product_name`, `name`, `category_id`. Пропусков получилось 423 - это все таргет 0.

## Создание таргета и фича-инжиниринг

### Создание таргета

После соединения таблиц появились пропуски. Там, где пропусков нет таргет будет 1, где есть - 0. Пропуски заполняем случайными данными из признаков.

In [34]:
df.loc[df['product_id'].isnull() == False,'match'] = 1
df.loc[df['product_id'].isnull() == True,'match'] = 0
df.isna().sum()

key               0
dealer_id       423
product_id      423
product_name      0
name            423
category_id     423
match             0
dtype: int64

In [35]:
temp = df[df['match'] == 0] # выделяем класс 0
temp1 = df[df['match'] == 1].sample(n=len(temp),random_state=RANDOM_STATE)# Выделяем случайные данные для класса 0
# Добавляем данные заказчика в таблицу с 0 классом.
for i in ['dealer_id', 'product_id', 'name', 'category_id']:
    temp[i] = list(temp1[i])
temp

key  dealer_id  product_id  \
44      90162708       17.0       332.0   
81    1001472231        5.0       128.0   
143     44182839       15.0       106.0   
148     33373363        8.0       168.0   
152     44056895       15.0       157.0   
...          ...        ...         ...   
3005   530170161        3.0       329.0   
3009   528911039       13.0        55.0   
3018   685030286       13.0       391.0   
3022  1090913025        5.0       276.0   
3023  1077090171        4.0        26.0   

                                           product_name  \
44    Спрей Prosept Cooky Spray для столешниц и рако...   
81    Огнебиозащитадлядревесины ProseptОГНЕБИОPROFII...   
143   Чистящее средство Universal Hard спрей для очи...   
148   Средство для удаления ржавчины Bath Acid с аро...   
152   Средство для мытья полов Multipower "После дож...   
...                                                 ...   
3005            Строительный антисептик  PROSEPT 5.6 кг   
3009  Герметик акриловый межшовный для деревянных ко...   
3018  Жидкое гель-мыло PROSEPT эконом-класса, без кр...   
3022  Герметик акриловый межшовный для деревянных ко...   
3023  Краска-грунт фасадная для плит OSB Proff 3 в 1...   

                                                   name  category_id  match  
44    Антисептик лессирующий Bi O LASUR / белый люкс...         31.0    0.0  
81    Средство для удаления клейкой ленты, клея, нак...         43.0    0.0  
143   Специальная соль для посудомоечных машин Splas...         42.0    0.0  
148   Пена для чистки ковров, мягкой мебели и тексти...         47.0    0.0  
152   Жидкое моющее средство для стирки белых и свет...         22.0    0.0  
...                                                 ...          ...    ...  
3005  Лак паркетный  PROSEPT, глянцевый,готовый сост...         31.0    0.0  
3009  Средство усиленного действия для удаления ржав...         52.0    0.0  
3018  Удалитель плесени FUNGI CLEAN концентрат 1:1 /...         35.0    0.0  
3022            Антисептик ECO ULTRA, коричневый / 10 л         20.0    0.0  
3023  Универсальный концентрат для мытья полов Multi...         50.0    0.0  

[423 rows x 7 columns]

In [36]:
dataframe_final = pd.concat([temp,(df.loc[df['match'] == 1])], ignore_index=True).drop(['key'], axis=1)

In [37]:
for i in ['dealer_id', 'product_id', 'match', 'category_id']:
    dataframe_final[i] = dataframe_final[i].astype(int)
dataframe_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3024 entries, 0 to 3023
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   dealer_id     3024 non-null   int32 
 1   product_id    3024 non-null   int32 
 2   product_name  3024 non-null   object
 3   name          3024 non-null   object
 4   category_id   3024 non-null   int32 
 5   match         3024 non-null   int32 
dtypes: int32(4), object(2)
memory usage: 94.6+ KB


In [38]:
dataframe_final

dealer_id  product_id  \
0            17         332   
1             5         128   
2            15         106   
3             8         168   
4            15         157   
...         ...         ...   
3019         18         258   
3020         18         299   
3021         18         452   
3022         18         416   
3023         18         405   

                                           product_name  \
0     Спрей Prosept Cooky Spray для столешниц и рако...   
1     Огнебиозащитадлядревесины ProseptОГНЕБИОPROFII...   
2     Чистящее средство Universal Hard спрей для очи...   
3     Средство для удаления ржавчины Bath Acid с аро...   
4     Средство для мытья полов Multipower "После дож...   
...                                                 ...   
3019  Строительный антисептик  PROSEPT Антисептик дл...   
3020  Огнебиозащита профессиональная для древесины P...   
3021  Огнебиозащита для древесины PROSEPT 2 группа, ...   
3022       Антисептик многофункциональный ФБС, ГОСТ 5 л   
3023       Средство для удаления ржавчины PROSEPT, 1 шт   

                                                   name  category_id  match  
0     Антисептик лессирующий Bi O LASUR / белый люкс...           31      0  
1     Средство для удаления клейкой ленты, клея, нак...           43      0  
2     Специальная соль для посудомоечных машин Splas...           42      0  
3     Пена для чистки ковров, мягкой мебели и тексти...           47      0  
4     Жидкое моющее средство для стирки белых и свет...           22      0  
...                                                 ...          ...    ...  
3019  Антисептик для внутренних работ PROSEPT INTERI...           20      1  
3020  Огнебиозащита ОГНЕБИО PROF 2 группа, красный г...           38      1  
3021  ОГНЕБИОЗАЩИТА для древесины 2 группа, красный ...           38      1  
3022     Антисептик многофункциональный ФБС, ГОСТ / 5 л           20      1  
3023  Удалитель ржавчины RUST REMOVER концентрат  1:...           35      1  

[3024 rows x 6 columns]

### Фича-инжиниринг

Добавление расстояния Левенштейна

In [39]:
def fuzzywuzzy_name(dataframe):
    return fuzz.token_set_ratio(dataframe['product_name'], dataframe['name'])

In [40]:
dataframe_final['fuzz'] = dataframe_final.apply(fuzzywuzzy_name, axis=1)

In [41]:
dataframe_final

dealer_id  product_id  \
0            17         332   
1             5         128   
2            15         106   
3             8         168   
4            15         157   
...         ...         ...   
3019         18         258   
3020         18         299   
3021         18         452   
3022         18         416   
3023         18         405   

                                           product_name  \
0     Спрей Prosept Cooky Spray для столешниц и рако...   
1     Огнебиозащитадлядревесины ProseptОГНЕБИОPROFII...   
2     Чистящее средство Universal Hard спрей для очи...   
3     Средство для удаления ржавчины Bath Acid с аро...   
4     Средство для мытья полов Multipower "После дож...   
...                                                 ...   
3019  Строительный антисептик  PROSEPT Антисептик дл...   
3020  Огнебиозащита профессиональная для древесины P...   
3021  Огнебиозащита для древесины PROSEPT 2 группа, ...   
3022       Антисептик многофункциональный ФБС, ГОСТ 5 л   
3023       Средство для удаления ржавчины PROSEPT, 1 шт   

                                                   name  category_id  match  \
0     Антисептик лессирующий Bi O LASUR / белый люкс...           31      0   
1     Средство для удаления клейкой ленты, клея, нак...           43      0   
2     Специальная соль для посудомоечных машин Splas...           42      0   
3     Пена для чистки ковров, мягкой мебели и тексти...           47      0   
4     Жидкое моющее средство для стирки белых и свет...           22      0   
...                                                 ...          ...    ...   
3019  Антисептик для внутренних работ PROSEPT INTERI...           20      1   
3020  Огнебиозащита ОГНЕБИО PROF 2 группа, красный г...           38      1   
3021  ОГНЕБИОЗАЩИТА для древесины 2 группа, красный ...           38      1   
3022     Антисептик многофункциональный ФБС, ГОСТ / 5 л           20      1   
3023  Удалитель ржавчины RUST REMOVER концентрат  1:...           35      1   

      fuzz  
0       32  
1       30  
2       39  
3       39  
4       43  
...    ...  
3019    99  
3020   100  
3021    98  
3022   100  
3023    54  

[3024 rows x 7 columns]

### Эмбеддинг

Будем использовать предобученный токенизатор и модель - RuBERT tiny. Объединим эти вектора скалярным произведением.

In [42]:
#загружаем токенизатор и модель
tokenizer = AutoTokenizer.from_pretrained("cointegrated/rubert-tiny")
model = AutoModel.from_pretrained("cointegrated/rubert-tiny")

In [43]:
%time
# Эмбеддинг для признака name
tokenized = dataframe_final['name'].apply((lambda x: tokenizer.encode(x, add_special_tokens=True)))
max_len = 0
for i in tokenized.values:
    if len(i) > max_len:
        max_len = len(i)

padded = np.array([i + [0]*(max_len-len(i)) for i in tokenized.values])
attention_mask = np.where(padded != 0, 1, 0)
input_ids = torch.tensor(padded)  
attention_mask = torch.tensor(attention_mask)

with torch.no_grad():
    last_hidden_states = model(input_ids, attention_mask=attention_mask)
vector_name = last_hidden_states[0][:,0,:].numpy()

Wall time: 0 ns


In [44]:
%time
# Эмбеддинг признака product_name
tokenized = dataframe_final['product_name'].apply((lambda x: tokenizer.encode(x, add_special_tokens=True)))
max_len = 0
for i in tokenized.values:
    if len(i) > max_len:
        max_len = len(i)

padded = np.array([i + [0]*(max_len-len(i)) for i in tokenized.values])
attention_mask = np.where(padded != 0, 1, 0)
input_ids = torch.tensor(padded)  
attention_mask = torch.tensor(attention_mask)

with torch.no_grad():
    last_hidden_states = model(input_ids, attention_mask=attention_mask)
vector_product_name = last_hidden_states[0][:,0,:].numpy()

Wall time: 0 ns


In [45]:
vector_name.shape[0]

3024

In [46]:
# Скалярное произведение
scalar = list()
for i in range(vector_name.shape[0]):
    scalar.append(np.dot(vector_product_name[i], vector_name[i]))

In [47]:
pd.DataFrame(scalar)

0
0      80.352013
1      77.235924
2      74.278069
3      69.504913
4      61.603527
...          ...
3019   93.345764
3020  101.697182
3021   99.547249
3022   97.399490
3023   97.046822

[3024 rows x 1 columns]

In [48]:
# Добавление признака с скалярным произведение
dataframe_final = pd.concat([pd.DataFrame(scalar, columns=['scalar']), dataframe_final], axis=1)
dataframe_final

scalar  dealer_id  product_id  \
0      80.352013         17         332   
1      77.235924          5         128   
2      74.278069         15         106   
3      69.504913          8         168   
4      61.603527         15         157   
...          ...        ...         ...   
3019   93.345764         18         258   
3020  101.697182         18         299   
3021   99.547249         18         452   
3022   97.399490         18         416   
3023   97.046822         18         405   

                                           product_name  \
0     Спрей Prosept Cooky Spray для столешниц и рако...   
1     Огнебиозащитадлядревесины ProseptОГНЕБИОPROFII...   
2     Чистящее средство Universal Hard спрей для очи...   
3     Средство для удаления ржавчины Bath Acid с аро...   
4     Средство для мытья полов Multipower "После дож...   
...                                                 ...   
3019  Строительный антисептик  PROSEPT Антисептик дл...   
3020  Огнебиозащита профессиональная для древесины P...   
3021  Огнебиозащита для древесины PROSEPT 2 группа, ...   
3022       Антисептик многофункциональный ФБС, ГОСТ 5 л   
3023       Средство для удаления ржавчины PROSEPT, 1 шт   

                                                   name  category_id  match  \
0     Антисептик лессирующий Bi O LASUR / белый люкс...           31      0   
1     Средство для удаления клейкой ленты, клея, нак...           43      0   
2     Специальная соль для посудомоечных машин Splas...           42      0   
3     Пена для чистки ковров, мягкой мебели и тексти...           47      0   
4     Жидкое моющее средство для стирки белых и свет...           22      0   
...                                                 ...          ...    ...   
3019  Антисептик для внутренних работ PROSEPT INTERI...           20      1   
3020  Огнебиозащита ОГНЕБИО PROF 2 группа, красный г...           38      1   
3021  ОГНЕБИОЗАЩИТА для древесины 2 группа, красный ...           38      1   
3022     Антисептик многофункциональный ФБС, ГОСТ / 5 л           20      1   
3023  Удалитель ржавчины RUST REMOVER концентрат  1:...           35      1   

      fuzz  
0       32  
1       30  
2       39  
3       39  
4       43  
...    ...  
3019    99  
3020   100  
3021    98  
3022   100  
3023    54  

[3024 rows x 8 columns]

In [49]:
# Удаление текста
dataframe_final.drop(['product_name', 'name'], inplace=True, axis=1)
dataframe_final

scalar  dealer_id  product_id  category_id  match  fuzz
0      80.352013         17         332           31      0    32
1      77.235924          5         128           43      0    30
2      74.278069         15         106           42      0    39
3      69.504913          8         168           47      0    39
4      61.603527         15         157           22      0    43
...          ...        ...         ...          ...    ...   ...
3019   93.345764         18         258           20      1    99
3020  101.697182         18         299           38      1   100
3021   99.547249         18         452           38      1    98
3022   97.399490         18         416           20      1   100
3023   97.046822         18         405           35      1    54

[3024 rows x 6 columns]

**Вывод:** В объединенной таблице полным строкам был присвоен таргет 1, а с пропусками - 0. Пропуски были заполнены рандомными значениями из соответственных признаков. `dealer_id`, `product_id`, `match`, `category_id` были переведены в int.

Была добавлена новая фича - расстояние Левенштейна.

Был произведен эмбеддинг на предобученной модели RuBERT tiny и скалярное произведение. Были удалены `product_name` и `name`. По итогу таблица состоит из:
- `scalar` - скалярное произведение двух эмбеддингов;
- `dealer_id` - идентификатор дилера;
- `product_ id` - идентификатор товара;
- `category_id` - категория товара;
- `match` - схожи ли названия между собой (таргет);
- `fuzz` - расстояние Левенштейна между названиями.

## Создание выборок и обучение модели

### Создание выборок

Две выборки - обучающая (80%) и тестовая (20%).

In [50]:
target = dataframe_final.match
features = dataframe_final.drop(['match'], axis=1)

In [51]:
target.value_counts()

1    2601
0     423
Name: match, dtype: int64

In [52]:
features_train, features_test, target_train, target_test = train_test_split(features, target, test_size=.2, random_state=RANDOM_STATE)

### Обучение модели

In [53]:
%%time
params = {
    'n_estimators': range(100, 1001, 100)
}
grid = GridSearchCV(LGBMClassifier(random_state=RANDOM_STATE), params, scoring='precision', n_jobs=-1)
grid.fit(features_train, target_train)
print(f'Лучший precision: {grid.best_score_}')

Лучший precision: 0.9832898467827466
Wall time: 3.62 s


In [54]:
grid.best_params_

{'n_estimators': 100}

**Вывод:** Было создано две выборки - обучающая (80%) и тестовая (20%). Есть дисбаланс классов, но модель показывает себя хорошо и с ним. Обучалась модель одна LGBMClassifier, она сразу показала себя прекрасно, нет смысла тестировать и другие. Гиперпараметры:
- n_estimators: 100.

Лучший precision - 0.98.

## Тестирование модели

In [55]:
print(f'Presicion на тестовой выборке: {precision_score(target_test, grid.predict(features_test))}')

Presicion на тестовой выборке: 0.9864864864864865


In [56]:
# Значимость признаков для модели
pd.DataFrame({'Value': grid.best_estimator_.feature_importances_, 'Feature': features.columns}).sort_values(by="Value", ascending=False)

Value      Feature
4    779         fuzz
2    735   product_id
0    686       scalar
1    519    dealer_id
3    281  category_id

**Вывод:** На тестовой выборке у модели не упало качество - 0.98. Важность признаков у модели распределено по всем.

## Вывод

В таблице `marketing_dealer`, всего 18 строк, дубликатов нет.

`dealer_price` содержала 20416 строк, пропуски в product_url, есть дубликаты (парсился несколько раз один и тот же продукт, но в разные дни), остаются признаки - `product_key` (для матчинга) и `product_name`. `product_key` переименую в `key`. После удаления дубликатов осталось 3024,

`product_dealer_key` имеет 1700 строк, дубликатов нет, на основе этой таблицы будет создан таргет - все эти связи имеют таргет 1, строки без пары будут иметь 0.

`product` содержит 496 строк, но есть пропуски в индексе - максимальный 513. Пропуски в `ean_13`, `name`, `cost`, `recommended_price`, `category_id`, `ozon_name`, `name_1c`, `wb_name`, `ozon_article`, `wb_article`, `ym_article`, `wb_article_td`. Оставлю признаки - `name` и `category_id`. Пропуски в `category_id` заполню заглушкой 0 и переведу в int. 

Таблицы `product`, `product_dealer_key` и `dealer_price` были объединены в одну, признаки - `key`, `dealer_id`, `product_id`, `product_name`, `name`, `category_id`. Пропусков получилось 423 - это все таргет 0.

В объединенной таблице полным строкам был присвоен таргет 1, а с пропусками - 0. Пропуски были заполнены рандомными значениями из соответственных признаков. `dealer_id`, `product_id`, `match`, `category_id` были переведены в int.

Была добавлена новая фича - расстояние Левенштейна.

Был произведен эмбеддинг на предобученной модели RuBERT tiny и скалярное произведение. Были удалены `product_name` и `name`. По итогу таблица состоит из:
- `scalar` - скалярное произведение двух эмбеддингов;
- `dealer_id` - идентификатор дилера;
- `product_ id` - идентификатор товара;
- `category_id` - категория товара;
- `match` - схожи ли названия между собой (таргет);
- `fuzz` - расстояние Левенштейна между названиями.

Было создано две выборки - обучающая (80%) и тестовая (20%). Есть дисбаланс классов, но модель показывает себя хорошо и с ним. Модель LGBMClassifier с гиперпараметры:
- n_estimators: 100.

Precision на тесте - 0.98.

## Выгрузка модели

In [57]:
pkl_filename = 'pickle_model.pkl'

with open(pkl_filename, 'wb') as file:
    pickle.dump(grid.best_estimator_, file)